### DISCUSSION TITLE
Create forms with query to read and create a email

### DISCUSSION THREAD

**Discussion by jaquelinedealmeida** [Budibase/budibase#13622] [Help] [answered]
1- In the first screen I need to verify if the email is in the database already. I am using a query to create a new register and the query adds only the field with a binding relation email. 
If the e-mail doesn't have in the database, it continues the navigation to the next screen.
If is in the database, stop the creation and send a warning. 
I put the actions in the button.
 
But when I test with an email registered already, I can navigate to the next screen.

How I can better this?

**Comment by mikesealey** [comment, marked as answer]
Hi @jaquelinedealmeida

There are two ways to go about this. Both look the same to the end user, but under the hood there's a few differences. To explore this, I created an app with a form-field that requests an email from the user, and `on click` of the button the app will check existing user emails. Using a `Navigate To` action and some JavaScript to return the path of the page depending on the response of the check

**Premium Users**
* If you're a premium user of Budibase, navigate to the Automation tab and create an App action. 
* Add a field to the app action called _email_, set to text.
* Add a _Query Rows_ step to the automation, select the table to query, set the limit to 1
* Define the filters: The column you wish to check (most likely email, in your case) EQUALS BINDING `{{ trigger.fields.email }}` (Your own variable names may differ slightly, use the bindings menu to assist).
* Add another step to your automation - this time _Collect Data_. Set it to collect the Step 1 Outputs rows (again, bindings menu will help).
* In the Design screen, at the point where you want to check if the email exists (ie a submit-button) you can then use actions to trigger that automation,  and a Navigate To action. (Again, your variables may differ, but...) my Navigate To action looks like this:

```JS
if ($("Action 1.Automation Result").value.length === 1) {
  return "/exists"
} else {
  return "/doesntexist"
}
```

The Automation Result is an object containing an array named Value - if there is an email address already in the table, it will appear here, giving the object a length of one, and therefor navigating to (in my case) /exists.

**Free Users**
* In the Data tab, create a new data source, and select REST API, then name it something useful like "Check Email Exists"
* Click Create New Query, select POST, and set the URL to location of your table - I'm using the Budibase sample data, so mine is https://budibase.app/api/public/v1/tables/ta_bb_employee/rows/search
* Add a binding "email" (use the default value here for testing/debugging, if you need to)
* Under "Headers" you'll need your API key and your App ID - you can find out where to get them both from in [the documentation here](https://docs.budibase.com/docs/public-api#how-to-get-your-api-key). You should add them under the keys "x-budibase-api-key" and "x-budibase-app-id" respectively. 
![image](https://github.com/Budibase/budibase/assets/39269380/ab7232a7-c5b7-48c1-81c5-e65fdfcdb33e)

* Under "Body" add the following as raw(JSON): 
```JS
{
  "query": {
		"equal": {
      "Email": "{{ Binding.email }}"
    }
  }
}
```

Where `Email` (capital E) is the column in the table you're querying, and `{{ Binding.email }}` is from the Bindings header.

* In the Design screen, at the point where you want to check if the email exists (ie a submit-button) you can then use actions to execute the query - you'll need to set your datasource and query to the REST API you've just created, and you should use the Bindings to capture the email from the form. 
* You should then use an Update State action to store the response of the query, followed by a Navigate To action which uses some JavaScript to decide where to go based on the value stored in state. 

```JS
if ($("Action 1.Query result").data.length === 1) {
  return "/exists"
} else {
  return "/doesntexist"
}
```


As always, your variables, bindings, and keys, will be unique to your project, so this answer should be used as a template rather than a solution. There's always more that can be said, so the [Budibase Discord Server](https://discord.gg/ZepTmGbtfF) is a useful place to be, and the [Budibase Documentation](https://docs.budibase.com/docs/what-is-budibase) is worth bookmarking to have open in a tab in the background for more guidance.

I hope this helps!

**Comment by jaquelinedealmeida** [comment]
Thank you. I am testing the aplication!

**Comment by mikesealey** [comment]
Hi @jaquelinedealmeida I'm sorry you're still having trouble. I've attached the export file of the app I used to explore this. I just used the sample data, and for ease I've included the table of emails on the home screen so you can copy/paste them from the table to the field for testing.

[13622-export-1715330961240.tar.gz](https://github.com/Budibase/budibase/files/15272699/13622-export-1715330961240.tar.gz)

You can read about how to import apps [in the documentation here.](https://docs.budibase.com/docs/export-and-import-apps)

**Comment by jaquelinedealmeida** [comment]
Thank you!! It's working! 
I change the condition!
```
if ($("Action 1.Query result").data.length === 0) {
  return "/new_page
} else {
  return "/same_page"
}
```

**Comment by jaquelinedealmeida** [comment]
Hi @mikesealey . I tested it again, but it's just going into the else.
Even if I change the order of the condition, the redirection of the screen to the unregistered email is always the same, as if it were in the database. I'm using a Rest API from nocodb and wondering if there could be any relation for that.

I still haven't been able to identify the error.
![Captura de tela 2024-05-13 173859](https://github.com/Budibase/budibase/assets/79230534/27e7f2a0-3d31-4db0-a003-02981fd8b1da)

![Captura de tela 2024-05-13 173906](https://github.com/Budibase/budibase/assets/79230534/8804229f-061f-483f-9ff3-38ec8e44bc61)
![Captura de tela 2024-05-13 173913](https://github.com/Budibase/budibase/assets/79230534/64d7df94-a4ca-4bdc-ad23-ee464f69be80)
![Captura de tela 2024-05-13 173926](https://github.com/Budibase/budibase/assets/79230534/bb8b7dcc-2bac-4c4a-8b16-5687b375a84e)

**Comment by mikesealey** [comment]
For debugging purposes, temporarily remove or disable the "Navigate To" action and add in a "Show Notification" which contains the $("Action1.Query result").

In it's current setup we're looking for an empty array, but the if-statement will take you down the else-path if you get literally anything else, so it could simply be that we've made an incorrect assumption earlier that is causing unexpected behaviour.

**Comment by jaquelinedealmeida** [comment]
I  fixed with `($("Action 1.Query result").data[0].list.length === 0)`

**Comment by mikesealey** [comment]
Perfect, I'm glad you got sorted!

**Reply by jaquelinedealmeida** [reply]
Thanks!
